src/background_client.cpp

```diff
@@ -0,0 +1,608 @@
+        {
+            auto const error_str = "Loading font from " + font_path.string() + " failed with error " + std::to_string(error);
+            mir::fatal_error(error_str.c_str());
+        }
+    }
+}
+
+TextRenderer::~TextRenderer()
+{
+    if (auto const error = FT_Done_Face(face))
+    {
+        mir::log_warning("Failed to uninitialize font face with error %d", error);
+    } 
+    face = nullptr;
+
+    if (auto const error = FT_Done_FreeType(library))
+    {
+        mir::log_warning("Failed to uninitialize FreeType with error %d", error);
+    }
+    library = nullptr;
+}
+
+void BackgroundClient::set_colour(std::string const& option, Colour& colour)
+{
+    uint32_t value;
+    std::stringstream interpreter{option};
+
+    if (interpreter >> std::hex >> value)
+    {
+        colour[0] = value & 0xFF;
+        colour[1] = (value >> 8) & 0xFF;
+        colour[2] = (value >> 16) & 0xFF;
+        // alpha should always be 255 in this instance
+        colour[3] = 255;
+    }
+    else
+    {
+        BOOST_THROW_EXCEPTION(std::runtime_error(
+            "Invalid colour (" + option + ") given in program argument"));
+    }
+}
+
+void BackgroundClient::set_wallpaper_top_colour(std::string const& option)
+{
+    set_colour(option, wallpaper_top_colour);
+}
+
+void BackgroundClient::set_wallpaper_bottom_colour(std::string const& option)
+{
+    set_colour(option, wallpaper_bottom_colour);
+}
+
+void BackgroundClient::set_crash_background_colour(std::string const& option)
+{
+    set_colour(option, crash_background_colour);
+}
+
+void BackgroundClient::set_crash_text_colour(std::string const& option)
+{
+    set_colour(option, crash_text_colour);
+}
+
+void BackgroundClient::set_diagnostic_path(std::string const& option)
+{
+    if (option.empty())
+    {
+        mir::log_info("No diagnostic file given. Crash reporting is disabled.");
+        return;
+    }
+    
+    auto option_path = Path(option);
+    if (boost::filesystem::exists(option_path.parent_path()))
+    {
+        diagnostic_path = option_path;
+    }
+    else
+    {
+        BOOST_THROW_EXCEPTION(std::runtime_error(
+            "Diagnostic path (" + option_path.parent_path().string() + ") does not exist"));
+    }
+}
+
+void BackgroundClient::render_background(
+    uint32_t width,
+    uint32_t height,
+    unsigned char* buffer,
+    Colour const& bottom_colour,
+    Colour const& top_colour)
+{
+    Colour new_pixel;
+
+    for (auto current_y = 0; current_y < height; current_y++)
+    {
+        // Render gradient
+        for (auto i = 0; i < 3; i++)
+        {
+            new_pixel[i] = ((current_y * bottom_colour[i] + (height - current_y) * top_colour[i]) / height);
+        }
+        new_pixel[3] = 0xFF;
+
+        // Copy new_pixel to buffer
+        auto const pixel_size = sizeof(new_pixel);
+        for (auto current_x = 0; current_x < pixel_size*width; current_x += pixel_size)
+        {
+            memcpy(buffer + current_x, new_pixel, pixel_size);
```

Comment:
Non blocking: you can avoid indexing as follows:
```suggestion
        for (auto current_x = buffer; current_x != buffer + pixel_size*width; current_x += pixel_size)
        {
            memcpy(current_x, new_pixel, pixel_size);
```

But the compiler probably rewrites for you anyway